src/compiler/transformers/esnext.ts

```diff
@@ -101,11 +114,204 @@ namespace ts {
                     return visitCatchClause(node as CatchClause);
+                case SyntaxKind.PropertyAccessExpression:
+                    return visitPropertyAccessExpression(node as PropertyAccessExpression);
+                case SyntaxKind.ClassDeclaration:
+                    return visitClassDeclaration(node as ClassDeclaration);
+                case SyntaxKind.ClassExpression:
+                    return visitClassExpression(node as ClassExpression);
                 default:
                     return visitEachChild(node, visitor, context);
             }
         }
 
+        function currentPrivateNameEnvironment() {
+            return privateNameEnvironmentStack[privateNameEnvironmentIndex];
+        }
+
+        function addPrivateName(name: PrivateName, initializer?: Expression) {
+            const environment = currentPrivateNameEnvironment();
+            const nameString = name.escapedText as string;
+            if (nameString in environment) {
+                throw new Error("Redeclaring private name " + nameString + ".");
+            }
+            const weakMap = createFileLevelUniqueName("_" + nameString.substring(1));
+            environment[nameString] = {
+                weakMap,
+                initializer
+            };
+            return weakMap;
+        }
+
+        function accessPrivateName(name: PrivateName) {
+            const environment = currentPrivateNameEnvironment();
+            const nameString = name.escapedText as string;
+            if (nameString in environment) {
+                return environment[nameString].weakMap;
+            }
+            // Undeclared private name.
+            return undefined;
+        }
+
+        function visitPropertyAccessExpression(node: PropertyAccessExpression): Expression {
+            if (isPrivateName(node.name)) {
+                const weakMapName = accessPrivateName(node.name);
+                if (!weakMapName) {
+                    return node;
+                }
+                return setOriginalNode(
+                    setTextRange(
+                        createClassPrivateFieldGetHelper(context, node.expression, weakMapName),
+                        /* location */ node
+                    ),
+                    node
+                );
+            }
+            return visitEachChild(node, visitor, context);
+        }
+
+        function visitorCollectPrivateNames(node: Node): VisitResult<Node> {
+            if (isPrivateNamedPropertyDeclaration(node)) {
+                addPrivateName(node.name);
+                return undefined;
+            }
+            // Don't collect private names from nested classes.
+            if (isClassLike(node)) {
+                return node;
+            }
+            return visitEachChild(node, visitorCollectPrivateNames, context);
+        }
+
+        function visitClassDeclaration(node: ClassDeclaration) {
+            startPrivateNameEnvironment();

```

Comment:
I don't think we need these double-visits: instead can just use the one visitor and use the `case`s in the one visitor to connect to what we want to do in the child nodes. This isn't just about performance: the way I'm proposing (if possible) seems more in line with the rest of the codebase and spirit of visitation.